bundles/org.eclipse.text/src/org/eclipse/jface/text/source/AnnotationModel.java

```diff
@@ -789,15 +799,31 @@ public void removeAllAnnotations() {
 	protected void removeAllAnnotations(boolean fireModelChanged) {
 		IAnnotationMap annotations= getAnnotationMap();
 		if (fDocument != null) {
-			Iterator<Annotation> e= getAnnotationMap().keySetIterator();
-			while (e.hasNext()) {
-				Annotation a= e.next();
-				Position p= annotations.get(a);
-				removePosition(fDocument, p);
-//				p.delete();
-				synchronized (getLockObject()) {
-					getAnnotationModelEvent().annotationRemoved(a, p);
+			if (!(annotations instanceof IAnnotationMap2)) {
+				Iterator<Annotation> e= getAnnotationMap().keySetIterator();
+				while (e.hasNext()) {
+					Annotation a= e.next();
+					Position p= annotations.get(a);
+					removePosition(fDocument, p);
+//					p.delete();
+					synchronized (getLockObject()) {
+						getAnnotationModelEvent().annotationRemoved(a, p);
+					}
 				}
+			} else {
+				// foreach is synchronized on annotations lock object
+				// if our lock is same, no extra synch block needed
+				boolean sameLock = getLockObject() == annotations.getLockObject();
+				annotations.forEach((a, p) -> {
+					removePosition(fDocument, p);
+					if (sameLock){						
```

Comment:
Note that getAnnotaitonModelEvent() also uses synchonrized(getLockObject()).
Iff acquiring a lock is a bottle-neck, you could even hoist the check to call different `forEach` lambdas:
```
if (sameLock) {
  var event = getAnnotationModelEvent();  
  annotations.forEach((a,p)->{
     removePosition(fDocument, p);
     event.annotationRemoved(a, p);
  });
} else {
  ..
}
```

Which makes me wonder whether we need IAnnotationMap2 or could just acquire the lock once and use the regular forEach(..)

Reply:
> Which makes me wonder whether we need IAnnotationMap2

I've created this interface because you mentioned that there are lot of existing code that might be using custom annotations and so could be affected by basic model changes. 

Originally I thought I can change the code to check if the lock is not null and use old code branch for null locks and new code for lock existing, but looking how lock object is used internally it seem to assume it is never null. So I left without a better idea with the "marker" interface approach.

But might be I misunderstood your concerns?

Reply:
> Iff acquiring a lock is a bottle-neck

Well, if we know we can avoid doing something in a loop, better avoid it. And yes, your proposal makes sense.

Reply:
Indeed I might have been not very clear. The concern was to not change the way the existing iterators work (based on lazy copying). I don't think we strictly need a new interface, but we have to guarantee that existing usage patterns of the IAnnotationMap do still work. 

Locally patched, I've this in production across multiple Eclipse versions and it doesn't seem to break anything:

```
AnnotationMap.java

	@Override
	public void forEach(BiConsumer<? super Annotation, ? super Position> action) {
		synchronized (fLockObject) {
			fInternalMap.forEach(action);
		}
	}
```
AnnotationModel.java
```
	@Override
	protected void removeAllAnnotations(boolean fireModelChanged) {
		synchronized (getLockObject()) {
			AnnotationModelEvent event = getAnnotationModelEvent();
			BidirectionalAnnotationMap map = getAnnotationMap();
			map.forEach((annotation, position) -> {
				removePosition(fDocument, position);
				event.annotationRemoved(annotation, position);
			});
			map.clear();
		}

		if (fireModelChanged) {
			fireModelChanged();
		}
	}
```